### ISSUE TITLE
DSSM not working well with Cross Entropy, but is working well with MSE

### ISSUE THREAD

**Issue by rajicon** [NTMC-Community/MatchZoo#641]
I'm experimenting with DSSM, and getting good results when I use MSE as the loss function.  However, when I use Cross Entropy (which I believe it is supposed to use), the results are very poor.  I was wondering if I was doing something wrong, so I'm posting my code here:

```python
train_raw = mz.pack(pd.read_csv('FILE', index_col=0))
test_raw = mz.pack(pd.read_csv('FILE', index_col=0))

train_raw.relation['label'] = train_raw.relation['label'].astype('float32')
test_raw.relation['label'] = test_raw.relation['label'].astype('float32')

preprocessor = mz.preprocessors.DSSMPreprocessor()

preprocessor.fit(train_raw)

train_preprocessed = preprocessor.transform(train_raw)
test_preprocessed = preprocessor.transform(test_raw)

model.params['input_shapes'] = preprocessor.context['input_shapes']
model.params['mlp_num_layers'] = 3
model.params['mlp_num_units'] = 300
model.params['mlp_num_fan_out'] = 128
model.params['mlp_activation_func'] = 'relu'
ranking_task = mz.tasks.Ranking(loss=mz.losses.RankCrossEntropyLoss(num_neg=5)) # line removed in mse
model.params['task'] = ranking_task  # line removed in mse
model.guess_and_fill_missing_params()

model.build()
model.compile()

x_train, y_train = train_preprocessed.unpack()
x_test, y_test = test_preprocessed.unpack()

history = model.fit(x_train, y_train, validation_data=(x_test, y_test), batch_size=120, epochs=20)

y_pred = model.predict(x_test)
y_pred = scipy.special.expit(y_pred) # line removed in mse
y_pred_cutoff = y_pred > .5

prec, recall, f1beta, support = precision_recall_fscore_support(y_test, y_pred_cutoff, pos_label=1, average='binary')
print(f1beta)
```

I'm not sure if I can post the data here, but it is an .csv with 5 columns, the number, id_left, text_left, id_right, text_right, and label.  Each matching row (with label 1) is followed by 5 rows of negative samples (with label 0).

Is there any mistake in the above code that leads to error in results?  Am I handling the negative samples correctly?

**Comment by bwanglzu** [comment]
@rajicon would you like to post your code somewhere using ipython notebook (so we can get a bit more info)?

**Comment by rajicon** [comment]
@bwanglzu   Sure, I will try to post something tonight or tomorrow.

**Comment by rajicon** [comment]
Here are the notebooks:

https://github.com/rajicon/matchzoo-experiments

The main notebook is matchzoo_toy_ce.ipynb , to be compared matchzoo_toy_mse.ipynb .  I made a toy example using the WikiQA dataset.  I took all the matches, and then generated 5 negative samples per pair.  The data is the .csv file.  

For this toy dataset, the results aren't good, but I'm more interested in whether my code is correct.  Am I using num_neg correctly?  Is there anything wrong with how I set the Ranking Task?  I think there may be problem with calculating probabilities after training.  I use the sigmoid function on the logits (as done in the notebook).  However, cross entropy in matchzoo uses softmax over all the negative samples, so this may not work as it should (I may post another issue about this).
